### ISSUE TITLE
faas-cli deploy works but result does not update - undeterministic behavior 

### ISSUE THREAD

**Issue by rohanmahapatra** [openfaas/faas#1701]
Hi.

I am building a Serverless function, specifically I am trying to perform ResNet50 inference as Serverless function. I change the handler.py, then do a faas-cli up (or even build, push, deploy). While invoking, most of the times, the result does not update and I see the old result. Sometimes, when I invoke the functionn 4-5 times, then the result changes and at times it does not change it all. I see indeterministic behavior at function invocation.

## Expected Behaviour
Ideally, the invocation should show me updated results reflecting my changes to the handler.py

## Are you a GitHub Sponsor (Yes/No?)
No

## List All Possible Solutions and Workarounds
I did try the following:
1. Used the --update and --replace flag as mentioned in this link: https://github.com/openfaas/faas/issues/4882. 
But it did not help. Rather, the invocation used to stall.

2. I removed the "/" part it from the "image" filed in the stack.yml file. I also tried removing the :latest tag but still I could not get it working.3. 3. 

3. I can run the docker standalone and it works fine.4. 

## Steps to Reproduce (for bugs)

Please find my handler.py contents below:
############### handler.py

import torch
from torch.autograd import Variable
## Sometimes, I get an error that the function could not find core 
from core import model

FUNCTION_ROOT = os.environ.get("function_root", "/home/app/function/")
# init model
#RNN = model.RNN(const.N_LETTERS, const.N_HIDDEN, const.N_CATEGORIES)
rnet = model.ResNet50()
rnet.load_state_dict(
    torch.load(str(PurePath(FUNCTION_ROOT, "data/resnet50.pt")))
)
def predict() -> List[Any]:
    rnet.eval()
    inputs = torch.randn(1,3,32,32)
    outputs = rnet(inputs)
    return outputs

def handle(req: bytes) -> str:
    """handle a request to the function
    Args:
        req (bytes): request body
    """
    if not req:
        return json.dumps({"error": "No input provided", "code": 400})
    output = predict()
    return "Success-v10"

############### stack.yml
version: 1.0
provider:
  name: openfaas
  gateway: http://127.0.0.1:8080
functions:
  resnet50-v1:
    lang: pydatascience
    handler: ./resnet50-v1
    image: rohanmahapatra/resnet50-v1:latest

## Your Environment
* FaaS-CLI version ( Full output from: `faas-cli version` ):
CLI:
 commit:  b1c09c0243f69990b6c81a17d7337f0fd23e7542
 version: 0.14.2
Gateway
 uri:     http://127.0.0.1:8080
 version: 0.21.3
 sha:     cdb6badddaede0f69ef8ce9cb4b7e25e876b75fa
Provider
 name:          faas-netes
 orchestration: kubernetes
 version:       0.14.2 
 sha:           54f00fd5acaceadca289e41d8d55360a0b4d5079

* Docker version `docker version` (e.g. Docker 17.0.05 ):
Client/Server both 20.10.7

* Are you using OpenFaaS on Kubernetes or [faasd](https://github.com/openfaas/faasd)?
Kubernetes (K3s)

* Operating System and version (e.g. Linux, Windows, MacOS):
Ubuntu 16.04.5 LTS

* Other diagnostic information / logs from [troubleshooting guide](https://docs.openfaas.com/deployment/troubleshooting)
I tried the below command but I kept seeing the same error even after I commented the lines on the handler.py
kubectl logs -n openfaas-fn deploy/resnet50-v1

I think my main question is why does not faas-cli up update my function. At times, I tried commenting out everything in the handler.py and just added a print statement. It still did not work or it used to work after a lot of tries.

Let me know if you need more information. 
Thanks

**Comment by alexellis** [comment]
Hi @rohanmahapatra 

OpenFaaS is deterministic in the way you describe, it will run what you provide.

Some users run into issues when they do not change their image tag between builds or have their openfaas chart misconfigured to use a pull policy of "IfNotPresent" instead of "Always"

Feel free to come along to the weekly community calls, you'll see the pinned issue in this repo.

I'm going to close as "not an issue" / as designed.

Hope this helps.

Alex

**Comment by alexellis** [comment]
/set title: Question about deploying updated functions

**Comment by japtain-cack** [comment]
I'm posting this here due to this being a top result when searching for "openfaas deploy not updating functions"

So I tag my images like such:
* 1
* 1.1
* 1.1.0

I typically pin my image to just the major version, which in this example would be 1. I'm also doing multi-arch builds, so I have to use the publish command. I believe this is causing the deploy to little to nothing in my case, since it builds/publishes an image to the docker hub and doesn't really touch the function deployment. Using `faas deploy --update=false --replace=true` results in a broken function, it deletes the function pods altogether and refuses to start them due to errors about a missing function in the gateway logs. After deleting the function and re-deploying it, that issue is resolved.

So, how do you get your pods to update when you aren't setting a new image version? With labels. It turns out, that setting function labels in the `stack.yml` will pass them to the pods. In the kubernetes documentation, it mentions adding a checksum to a configmap to force new config files to trigger an update. So, I applied this concept here and it works great.

Simply get a sha of all the files that you care about, I use `src/` for my function files. So I grab a sha1 hash of all the files under `/src/` and then append that to the function labels.

`make.sh`
```shell
#!/usr/bin/env bash

function publish(){
  faas publish --platforms linux/arm64,linux/amd64
}

function deploy(){
  FILE_CHECKSUMS=()
  for file in src/*; do
    SHASUM=(`shasum ${file}`)
    CHECKSUM=${SHASUM[0]}
    CHECKFILE=${SHASUM[1]}
    FILE_CHECKSUMS+=("${CHECKFILE}: ${CHECKSUM}")
  done
  echo Determining sha1 of files: ${FILE_CHECKSUMS[@]}
  CHECKSUMS=$(echo -e "${FILE_CHECKSUMS[@]}" | openssl sha1 | cut -d " " -f 2)
  echo Final checksum of all files: $CHECKSUMS
  CHECKSUMS=$CHECKSUMS faas deploy
}

case $1 in
  publish)
    publish
    ;;

  deploy)
    deploy
    ;;

  all)
    publish && deploy
    ;;

  *)
  echo Usage: make [options]
  echo
  echo Options:
  echo publish
  echo deploy
  echo all
esac
```

`stack.yml`
```yaml
version: 1.0
provider:
  name: openfaas
  gateway: https://fn.example.com
functions:
  ip:
    labels:
      com.openfaas.scale.min: 2
      src/sha1: ${CHECKSUMS}
    image: ip:1
    lang: python3-flask-debian
    handler: ./src
configuration:
  templates:
    - name: python3-flask
      source: https://github.com/openfaas/python-flask-template
```
